execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,498 @@
+                        "Key not found while attempting to apply transform"
+                    );
+                    return Err(CommitError::KeyNotFound(key).into());
+                }
+                (Some(current_value), transform) => match transform.apply(current_value.clone()) {
+                    Ok(updated_value) => updated_value,
+                    Err(err) => {
+                        error!(?key, ?err, "Key found, but could not apply transform");
+                        return Err(CommitError::TransformError(err).into());
+                    }
+                },
+            };
+
+            self.cache.write().unwrap().insert(key, value);
+        }
+        Ok(state_hash)
+    }
+}
+
+impl StateProvider for ScratchGlobalState {
+    type Error = error::Error;
+
+    type Reader = ScratchGlobalStateView;
+
+    fn checkout(&self, state_hash: Digest) -> Result<Option<Self::Reader>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let maybe_root: Option<Trie<Key, StoredValue>> = self.trie_store.get(&txn, &state_hash)?;
+        let maybe_state = maybe_root.map(|_| ScratchGlobalStateView {
+            cache: Arc::clone(&self.cache),
+            environment: Arc::clone(&self.environment),
+            store: Arc::clone(&self.trie_store),
+            root_hash: state_hash,
+        });
+        txn.commit()?;
+        Ok(maybe_state)
+    }
+
+    fn empty_root(&self) -> Digest {
+        self.empty_root_hash
+    }
+
+    fn get_trie(
+        &self,
+        _correlation_id: CorrelationId,
+        trie_key: &Digest,
+    ) -> Result<Option<Trie<Key, StoredValue>>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let ret: Option<Trie<Key, StoredValue>> = self.trie_store.get(&txn, trie_key)?;
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn put_trie(
+        &self,
+        correlation_id: CorrelationId,
+        trie: &Trie<Key, StoredValue>,
+    ) -> Result<Digest, Self::Error> {
+        let mut txn = self.environment.create_read_write_txn()?;
+        let trie_hash = put_trie::<
+            Key,
+            StoredValue,
+            lmdb::RwTransaction,
+            LmdbTrieStore,
+            Self::Error,
+        >(correlation_id, &mut txn, &self.trie_store, trie)?;
+        txn.commit()?;
+        Ok(trie_hash)
+    }
+
+    /// Finds all of the keys of missing descendant `Trie<K,V>` values
+    fn missing_trie_keys(
+        &self,
+        correlation_id: CorrelationId,
+        trie_keys: Vec<Digest>,
+    ) -> Result<Vec<Digest>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let missing_descendants =
+            missing_trie_keys::<Key, StoredValue, lmdb::RoTransaction, LmdbTrieStore, Self::Error>(
+                correlation_id,
+                &txn,
+                self.trie_store.deref(),
+                trie_keys,
+            )?;
+        txn.commit()?;
+        Ok(missing_descendants)
+    }
+}
```

Comment:
If `ScratchGlobalState` struct contained `LmdbGlobalState` then all of these could just forward to the underlying DB and only `StateReader::read` and `CommitProvider::commit` would reimplement their logic on top of it.